### ISSUE TITLE
Interceptor doing the file ops administration itself over shm

### ISSUE THREAD

**Issue by egmontkob-work** [firebuild/firebuild#618]
Instead of our current model, where the interceptor somehow notifies the supervisor about all kinds of events, and the supervisor does the administration...

For simple file operations (opening, closing etc.) maybe the interceptor could do all of this itself, without having to talk to the supervisor, thus saving the cost of communication, context switches etc.

The supervisor, and the communication channel would remain in the game for more complex events (fork/exec/alike).

---

This would need that some of our data structures (most notably: the global FileNames, and the per-process FileUsages) are over shm.

Some possible technical solutions:
- https://github.com/simonhf/sharedhashfile
- https://www.boost.org/doc/libs/1_50_0/doc/html/interprocess/allocators_containers.html#interprocess.allocators_containers.additional_containers.unordered (it's C++, I don't want to bring it back to the interceptor)
- something we build :)

---

The way I currently imagine this feature:

There'd be one shm shared across all processes, containing the global FileNames. Access to this would probably be guarded by a pthread_rwlock rather than a mutex, I presume that's faster because there'd be way more read than write requests (think about the system-wide header files).

There'd be one shm for each ExecedProcess that is still shortcuttable. The interceptor would map its own area, as well as the areas for its ancestors all the way up to the top process. Whenever there's a simple file operation happening, the process would grab the mutex on its own area, do the administration, release the mutex, grab the mutex of the parent, do the administration there etc., bubbling all the way up.

Automatic growing of the backed area is easly doable like it's already done in the shm branch.

We'd have a fix sized header in each of our shms, e.g. for the process we could denote there if it's shortcuttable etc. Then we would only need to have one hashmap per shm (either the FileNames or the FileUsages). If I'm right with this, maybe implementing our own hashmap solution is not that hard either.

Of course file contents hash computation would then also reside in the interceptor (or in both components). However, due to the global shm, we could prevent computing it over and over again, we'd still only compute it once (or, as opposed to our current design, a very few times if multiple processes open the same file for the first time in parallel - I think the cost of it is negligible.)

More complex actions, like system/popen will still use the communication channel. So sometimes the supervisor will need to do the administration instead of the interceptor. It would use the same code, including the same rwlocking/mutexing.

---

This is going to be quite some work to implement. We should think about how we can come up with a reasonable estimate on the expected performance improvement, without doing all this work, if possible.
